api/db/helper.go

```diff
@@ -0,0 +1,224 @@
+func (c *Client) Get(result interface{},
+	query *builder.Builder) error {
+
+	sql, args, err := prepareQuery(query)
+	if err != nil {
+		log.Debugf("ERROR Preparing query: %s", err)
+		return err
+	}
+	return sqlx.Get(c.ex, result, sql, args...)
+}
+
+func prepareQuery(query *builder.Builder) (string, []interface{}, error) {
+	sql, args, err := query.ToSQL()
+	if err != nil {
+		return "", nil, err
+	}
+	sql, err = builder.ConvertPlaceholder(sql, "$")
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sql, args, nil
+}
+
+// SelectWithCount executes the query and scans the results into the passed interface
+// Additionally it fetches the rowcount and adds Pagination
+func (c *Client) SelectWithCount(
+	result interface{},
+	query *builder.Builder,
+	pageNumber, perPage int) (int, error) {
+
+	// Check that the result slice sent in is an empty slice
+	//		created via ```name = New(type)``` or ``` name := type{}```
+	//		has a pointer to an empty array
+	// rather than a nil slice
+	//		created via ```var name type```
+	//		has no pointer
+	// as a nil array is json encoded as null rather than []
+	if reflect.Indirect(reflect.ValueOf(result)).IsNil() {
+		return 0, errors.New("Result list must be empty and not nil")
+	}
+
+	// copy the struct so we can replace the select fields
+	countQuery := &builder.Builder{}
+	*countQuery = *query
+	countQuery.Select("count(*)").OrderBy("")
+
+	sql, args, err := countQuery.ToSQL()
+	if err != nil {
+		return 0, err
+	}
+	sql, err = builder.ConvertPlaceholder(sql, "$")
+	if err != nil {
+		return 0, err
+	}
+	log.Debugf("SQL: %s : %+v", sql, args)
+
+	total := 0
+	err = sqlx.Get(c.ex, &total, sql, args...)
+	if err != nil {
+		log.Debugf("ERROR Getting Totals: %s", err)
+		return 0, err
+	}
+
+	sql, args, err = query.ToSQL()
+	if err != nil {
+		log.Debugf("ERROR Getting Rows: %s", err)
+		return 0, err
+	}
+	sql, err = builder.ConvertPlaceholder(sql, "$")
+	if err != nil {
+		return 0, err
+	}
+
+	sql = c.paginate(sql, pageNumber, perPage)
+	return total, sqlx.Select(c.ex, result, sql, args...)
+}
+
+// SelectWithCountSQL executes the query (raw sql)  and scans the results into the passed interface
+// Additionally it fetches the rowcount and adds Pagination
+func (c *Client) SelectWithCountSQL(result interface{}, query string, params []interface{},
+	orderBy string, pageNumber, perPage int) (int, error) {
+
+	var total int
+	err := sqlx.Get(c.ex, &total, "SELECT count(*) FROM ("+query+") AS orig", params...) // nolint: gosec
+	if err != nil {
+		return 0, err
+	}
+
+	if orderBy != "" {
+		query += " ORDER BY " + orderBy
+	}
+	query = c.paginate(query, pageNumber, perPage)
+	err = sqlx.Select(c.ex, result, query, params...)
+	if err != nil {
+		return 0, err
+	}
+
+	return total, nil
+}
+
+func (c *Client) paginate(query string, pageNumber int, perPage int) string {
+	if pageNumber > 0 && perPage > 0 {
+		// TODO Change to use placeholders, go-xorm/builder should support it soon
```

Comment:
needs to be addressed before final product launch

Reply:
It will be changed once it's oficially supported.

Reply:
all okay, long term to do, will keep up to date to implement when supported